hazelcast/src/main/java/com/hazelcast/map/impl/MapEntries.java

```diff
@@ -35,31 +35,139 @@
  */
 public final class MapEntries implements IdentifiedDataSerializable {
 
+    private static interface ValueWrapper {
```

Comment:
What about changing the wrapper to also include the key? This would allow collapsing the keys and values lists into a single list of entries. It also makes the putAllTo* methods more efficient as only a single Iterator needs to be created/managed.
Do you have some other specific change in mind?
Do you have benchmarks to measure the impacts you are concerned about?

My guess is that changes to initialize collections to correct sizes, such as done https://github.com/hazelcast/hazelcast/pull/9597 and in this PR reduce more object creation than what is introduced here.

I have added changes to use a single Entry class in MapEntries, which avoids polymorphism.